### ISSUE TITLE
Contrast works, brightness does not. Anything I can do?

### ISSUE THREAD

**Issue by agileadam** [kfix/ddcctl#9]
Hello,

I was very excited to see this project. Awesome job!
I can control the brightness of my iMac display. I can control the contrast of my two ASUS VE248 displays. Unfortunately, the brightness doesn't seem to be controllable. Is there anything I can do to get that working, or is the hardware simply not able to support it?

Thanks for the great tool!

**Comment by agileadam** [comment]
I'm not sure if it's helpful, but here's the output when I query contrast, volume, and brightness:

```
~/3rdparty/ddcctl $ ddcctl -d 2 -c ?
2015-11-20 08:40:17.600 ddcctl[19601:6160625] D: NSScreen #69992768 (2560x1440) DPI is 109.00
2015-11-20 08:40:17.600 ddcctl[19601:6160625] D: NSScreen #441001217 (1920x1080) DPI is 92.00
2015-11-20 08:40:17.601 ddcctl[19601:6160625] D: NSScreen #441001218 (1920x1080) DPI is 92.00
2015-11-20 08:40:17.601 ddcctl[19601:6160625] I: found 3 displays
2015-11-20 08:40:17.601 ddcctl[19601:6160625] I: polling display 2's EDID
2015-11-20 08:40:17.627 ddcctl[19601:6160625] I: got edid.name: VE248
2015-11-20 08:40:17.627 ddcctl[19601:6160625] I: got edid.serial: A9LMQS013537
2015-11-20 08:40:17.627 ddcctl[19601:6160625] D: command arg-pair: d: 2
2015-11-20 08:40:17.628 ddcctl[19601:6160625] D: command arg-pair: c: ?
2015-11-20 08:40:17.628 ddcctl[19601:6160625] D: querying VCP control: #18 =?
2015-11-20 08:40:17.733 ddcctl[19601:6160625] I: VCP control #18 = current: 72, max: 100

~/3rdparty/ddcctl $ ddcctl -d 2 -v ?
2015-11-20 08:40:20.682 ddcctl[19602:6160715] D: NSScreen #69992768 (2560x1440) DPI is 109.00
2015-11-20 08:40:20.683 ddcctl[19602:6160715] D: NSScreen #441001217 (1920x1080) DPI is 92.00
2015-11-20 08:40:20.684 ddcctl[19602:6160715] D: NSScreen #441001218 (1920x1080) DPI is 92.00
2015-11-20 08:40:20.684 ddcctl[19602:6160715] I: found 3 displays
2015-11-20 08:40:20.684 ddcctl[19602:6160715] I: polling display 2's EDID
2015-11-20 08:40:20.710 ddcctl[19602:6160715] I: got edid.name: VE248
2015-11-20 08:40:20.711 ddcctl[19602:6160715] I: got edid.serial: A9LMQS013537
2015-11-20 08:40:20.711 ddcctl[19602:6160715] D: command arg-pair: d: 2
2015-11-20 08:40:20.711 ddcctl[19602:6160715] D: command arg-pair: v: ?
2015-11-20 08:40:20.711 ddcctl[19602:6160715] D: querying VCP control: #0 =?
2015-11-20 08:40:20.816 ddcctl[19602:6160715] E: DDC send command failed!
2015-11-20 08:40:20.816 ddcctl[19602:6160715] E: VCP control #0 = current: 128, max: 128

~/3rdparty/ddcctl $ ddcctl -d 2 -b ?
2015-11-20 08:40:25.363 ddcctl[19603:6160783] D: NSScreen #69992768 (2560x1440) DPI is 109.00
2015-11-20 08:40:25.364 ddcctl[19603:6160783] D: NSScreen #441001217 (1920x1080) DPI is 92.00
2015-11-20 08:40:25.364 ddcctl[19603:6160783] D: NSScreen #441001218 (1920x1080) DPI is 92.00
2015-11-20 08:40:25.364 ddcctl[19603:6160783] I: found 3 displays
2015-11-20 08:40:25.364 ddcctl[19603:6160783] I: polling display 2's EDID
2015-11-20 08:40:25.392 ddcctl[19603:6160783] I: got edid.name: VE248
2015-11-20 08:40:25.393 ddcctl[19603:6160783] I: got edid.serial: A9LMQS013537
2015-11-20 08:40:25.393 ddcctl[19603:6160783] D: command arg-pair: d: 2
2015-11-20 08:40:25.393 ddcctl[19603:6160783] D: command arg-pair: b: ?
2015-11-20 08:40:25.393 ddcctl[19603:6160783] D: querying VCP control: #16 =?
2015-11-20 08:40:25.498 ddcctl[19603:6160783] I: VCP control #16 = current: 0, max: 100
```

**Comment by kfix** [comment]
> I can control the brightness of my iMac display.   

Care to share your iMac's model# and a `ddcctl -d 1` of it?

> I: VCP control #16 = current: 0, max: 100  

-b having a value of 0 is unusual. What happens when you try to set it to something?

> Is there anything I can do to get that working, or is the hardware simply not able to support it?  

there is an undocumented dump flag (`-D`) that will try polling all 256 possible DDC control id's.
Perhaps brightness is hiding under a different ID for this monitor. 
Try dumping  once (`-d 2 -D`), change the brightness with the physical controls, and dumping again to find a difference.

**Comment by agileadam** [comment]
Wow, you responded quickly!

Here's the output from ddcctl -d 1:
```
2015-11-20 16:08:39.568 ddcctl[20545:7300231] D: NSScreen #69992768 (2560x1440) DPI is 109.00
2015-11-20 16:08:39.569 ddcctl[20545:7300231] D: NSScreen #441001217 (1920x1080) DPI is 92.00
2015-11-20 16:08:39.569 ddcctl[20545:7300231] D: NSScreen #441001218 (1920x1080) DPI is 92.00
2015-11-20 16:08:39.569 ddcctl[20545:7300231] I: found 3 displays
2015-11-20 16:08:39.570 ddcctl[20545:7300231] I: polling display 1's EDID
2015-11-20 16:08:39.579 ddcctl[20545:7300231] I: got edid.name: iMac
2015-11-20 16:08:39.580 ddcctl[20545:7300231] D: command arg-pair: d: 1
```
My iMac is a 27-inch, Late 2013.
My external monitors are hooked up via mini display port to DVI.

The undocumented -D dump flag doesn't seem to be returning anything on all 3 screens.

```
~/3rdparty/ddcctl $ ddcctl -d 2 -D
2015-11-20 16:16:48.627 ddcctl[20575:7332383] D: NSScreen #69992768 (2560x1440) DPI is 109.00
2015-11-20 16:16:48.628 ddcctl[20575:7332383] D: NSScreen #441001217 (1920x1080) DPI is 92.00
2015-11-20 16:16:48.629 ddcctl[20575:7332383] D: NSScreen #441001218 (1920x1080) DPI is 92.00
2015-11-20 16:16:48.629 ddcctl[20575:7332383] I: found 3 displays
2015-11-20 16:16:48.629 ddcctl[20575:7332383] I: polling display 2's EDID
2015-11-20 16:16:48.656 ddcctl[20575:7332383] I: got edid.name: VE248
2015-11-20 16:16:48.656 ddcctl[20575:7332383] I: got edid.serial: A9LMQS013537
2015-11-20 16:16:48.656 ddcctl[20575:7332383] D: command arg-pair: d: 2
```

**Comment by kfix** [comment]
whoops...try `ddccdl -d 2 -D 1`

**Comment by agileadam** [comment]
Alright, I'm embarrassed. I executed once before adjusting monitor. It ran great with `ddcctl -d 2 -D 1` -- I dumped the output into Vim, cleaned up the time-related info at beginning of each line, then went to change my brightness via buttons on screen... realized my monitor was on a mode where brightness was disabled! DOH.

`ddcctl -d 2 -b ?` returns the correct number now!

This tool is going to make my life so much easier! I've been trying screen dimming apps like Flux and myPoint Shade and EasyOnTheEyes but overlay-based dimming stinks. I take a lot of screenshots and unless I turn that off the screenshots contain the overlay too. 

I'm going to build a simple GUI (or maybe even just use Keyboard Maestro) to increase/decrease all screens' brightness in sync.

Thanks so much!
